### PR TITLE
[FSDP][Perf] Avoid post-backward pad

### PR DESCRIPTION
**Pull request by awgu**
**Overview**
This PR adds an additional code path for when the `FlatParameter` must be padded to achieve an unsharded size divisible by the world size to enable the performant `_all_gather_base()` and `_reduce_scatter_base()`.
- In the padded case, we register a new pre-backward hook on the `FlatParameter` itself that _pre-allocates_ the padded unsharded gradient as a zero tensor to avoid doing so in the post-backward hook, which would (1) allocate unsharded memory in the post-backward stream and (2) require a D2D copy, which may incur nontrivial overhead and block the subsequent reduce-scatter.
    - The newly-computed unsharded gradient is added to the pre-allocated zero gradient. From a running time perspective, we tradeoff a GPU `memcpy` kernel with a GPU `add` kernel, so the running time may not decrease.
    - The `handle.prepare_gradient_for_backward()` call is refactored to be in this new pre-backward hook `_flat_param_pre_backward_hook()`. This is to delay the pre-allocation as much as possible. The new hook is guarded by `handle._ran_flat_param_pre_backward_hook` because otherwise the hook may run multiple times per backward. It is reset to `False` in the post-backward hook.
    - This approach leverages the private API `torch._C._will_engine_execute_node(flat_param_acc_grad)` to check if a `FlatParameter` will compute a gradient in the current `backward()`. As a part of this, `FlatParameter`s that are not involved in the current `backward()` are not all-gathered.
- We do not pre-allocate for the non-padded case because that would allocate an extra tensor compared to the existing behavior.
- For both the padded and non-padded cases, this PR changes the sharded gradient allocation to use `torch.empty()` instead of `torch.zeros()` to avoid a `fill` kernel.
- This PR renames `param` to `flat_param` in the post-backward hook for clarity.

**Example: RegNet-3B**
For RegNet-3B using an auto wrap policy that wraps `RegNetYLayer` and world size of 8, there are both `FlatParameter`s that need padding and those that do not.

With this PR (using a batch size of 48, `BACKWARD_PRE` prefetching, FP16 mixed precision, and activation checkpointing):
- The number of `memcpy` kernels per training iteration decreases from 234 to 6.
- [`use_orig_params=False`] The peak reserved memory decreases from 16.625 GB to 16.551 GB.
- [`use_orig_params=False`] The peak active memory decreases from 10.500 GB to 10.363 GB.
- [`use_orig_params=True`] The peak reserved memory decreases from 14.051 GB to 13.977 GB.
- [`use_orig_params=True`] The peak active memory decreases from 10.501 GB to 10.363 GB.

<details>
    <summary>Example: 2nd ReLU Backward</summary>

As an example, we examine the 2nd ReLU backward. First, we look at the trace for the existing code:
Screen Shot 2022-11-07 at 9 23 31 AM
Screen Shot 2022-11-07 at 9 23 22 AM
We see above the `memcpy` kernel. Now, we look at the trace for the new code:
Screen Shot 2022-11-07 at 9 23 38 AM
Screen Shot 2022-11-07 at 9 25 39 AM
We see an `add` kernel above but no more `memcpy` kernel. I am not sure why exactly, but this somehow allows the reduce-scatter to run earlier. (The preceding `BUnaryFunctor` kernel right before the reduce-scatter somehow runs earlier. This behavior is consistent across multiple runs.)

</details>

<details>
<summary> _flat_param_pre_backward_hook() </summary>

This allocates a zero tensor for the padded unsharded gradient, and it runs after the normal pre-backward hook.
Screen Shot 2022-11-07 at 8 30 24 AM

</details>

### CODE REVIEW

**Comment by awgu** [comment]
@soulitzer Is it okay for FSDP to depend on `torch._C._will_engine_execute_node()`?

**Comment by awgu** [comment]
Closing this to reopen with ghstack
